### ISSUE TITLE
[FEAT] Improve kornia.color

### ISSUE THREAD

**Issue by edgarriba** [kornia/kornia#145]
I open this since we might want to implement the following operators in `kornia.color` module. The motivation is that torchvision only supports this for PIL images, we want to give support for torch.Tensor:

Assigned -> @priba

- [x]  `kornia.color.adjust_brightness` https://github.com/arraiyopensource/kornia/pull/195
- [x] `kornia.color.adjust_contrast` [torchvision ref](https://pytorch.org/docs/stable/torchvision/transforms.html#torchvision.transforms.functional.adjust_contrast)
- [x] `kornia.color.adjust_hue` [torchvision ref](https://pytorch.org/docs/stable/torchvision/transforms.html#torchvision.transforms.functional.adjust_hue)
- [x] `kornia.color.adjust_saturation` [torchvision ref](https://pytorch.org/docs/stable/torchvision/transforms.html#torchvision.transforms.functional.adjust_saturation)
- [x]  `kornia.color.adjust_gamma` [torchvision ref](https://pytorch.org/docs/stable/torchvision/transforms.html#torchvision.transforms.functional.adjust_gamma)

Extra:
- [ ] `kornia.color.zca_whitening` [numpy ref](https://github.com/mwv/zca/blob/master/zca/zca.py#L32)

**Comment by bhack** [comment]
I could be nice to have other color operators like DeepMind/Tensorflow:

https://github.com/deepmind/multidim-image-augmentation/blob/master/doc/color_augmentation_colab.md

**Comment by varunagrawal** [comment]
@edgarriba quick question: does it make sense to have the operations have batch options?
I personally feel this makes the code more complex for a very small use case. In the most likely cases, the user will perform the transformation in the dataloader before the batching takes place.

**Comment by ducha-aiki** [comment]
Sure it makes sense. You can do then lots of color transforms, then concat and learn The color. 
https://github.com/ducha-aiki/caffenet-benchmark/blob/master/Colorspace.md

There was also recent paper about it. 

Or - you may want to learn color-invariant transformation. One of the ways to do it - make that transform differentiable, like for the geometry (spatial transformers)

**Comment by edgarriba** [comment]
Or maybe the "standard" approach for data augmentation is incorrect and the way to go is doing the transforms once the batch is created directly in the you

**Comment by varunagrawal** [comment]
Alright.

However, this issue needs to be better defined. The transformations requested make use of the properties of various color spaces, and this requires the input tensor to be `uint8` for many of those color space transforms to make sense. This is also how PIL does it.

However, #195 seems to accept tensors between 0 and 1, and this begs the question, whether is this valid or should we restrict users to input tensors only within `[0, 255]`? Also, what about the number of channels? Should there be a hard check for the number of channels to be 3 or 4 depending on the presence of an alpha mask? It doesn't make sense to allow an arbitrary number of channels.

**Comment by priba** [comment]
Hi,

At the moment I considered the input to be between 0 and 1. When I thought about it, it makes more sense from my point of view. Shall we change or allow it to be between [0, 255]?

Also, I am a bit confused when I compare the brightness/contrast adjust between PIL (torchvision) and OpenCV. In [OpenCV](https://docs.opencv.org/3.4.3/d3/dc1/tutorial_basic_linear_transform.html), brightness is defined as an additive operation whereas in [PIL](https://github.com/python-pillow/Pillow/blob/master/src/PIL/ImageEnhance.py#L74) it is the interpolation by a factor between a 0 image and the original image, hence, it is a multiplicative operation. Any idea about this?

**Comment by varunagrawal** [comment]
What would your point of view be? Most image manipulation programs use `uint8` types with the values being `[0-255]` since the conversion operations for different color spaces are defined with respect to that (though one can argue mathematically that 0-255 can be scaled to 0-1, but I am not sure how that would affect clamping of values , which it technically shouldn't).

Hence my argument for fleshing this issue out in more detail so we know exactly what the requirements are. Since the issue mentions that the operations should be akin to torchvision, this implies we should be working in the `0-255` range.

**Comment by priba** [comment]
As kornia is a differentiable computer vision library I understand that these modules can be used in any stage of the model. Hence, we will usually deal with real values rather than `uint8`. That's why I considered the input to be between 0 and 1.

**Comment by edgarriba** [comment]
@priba you are right, the assumption is correct, we are not a standard vision library and we might want to represent images from the point of view of feature tensors.
@varunagrawal It's true that we might want to mimic some of the `torchvision` interfaces however the goal of the library is for training, meaning that the operators don't have to be limited to consume only `uint8` images. We should expect feature tensors (image intensities can be seen as features too). I'd would stick to 0-1 range as input for the operators in this module. If it's well documented it shouldn't be a problem.

@priba are you continuing with the missing items ? maybe @varunagrawal can give you a hand.

BTW, if anyone is interested in helping @ShahriarSS, he's trying stuff related to this module as a starting point to port code into c++ and python to improve performance.
https://github.com/ShahriarSS/kornia/tree/c++-bindings

**Comment by priba** [comment]
@edgarriba I am a bit confused with the differences between PIL (torchvision) and OpenCV for brightness and contrast. Once this is clear I can continue with the missing items.

**Comment by edgarriba** [comment]
@priba what are the actual differences? can we bootstrap this somehow ? /cc @dvd42 can you help with that task?

**Comment by edgarriba** [comment]
closing this. Will open new one with the whitening function
